# HG changeset patch
# User adf88@interia.pl
# Date 1482656406 -3600
#      Sun Dec 25 10:00:06 2016 +0100
# Branch trunk
# Node ID f14ce943829b0182a274038e5dec317a08b8d98e
# Parent  fd244df3071e74c5f400d9a33256d1806c2a19e1
[mq]: fix-clipping-by-widget-clip

diff -r fd244df3071e -r f14ce943829b src/widget.cpp
--- a/src/widget.cpp	Sun Dec 25 09:59:20 2016 +0100
+++ b/src/widget.cpp	Sun Dec 25 10:00:06 2016 +0100
@@ -2393,15 +2393,22 @@
 {
 	if (this->current_x == 0 || this->current_y == 0) return;
 
+	/* Setup a clipping rectangle... */
+	DrawPixelInfo new_dpi;
+	if (!FillDrawPixelInfo(&new_dpi, this->pos_x, this->pos_y, this->current_x, this->current_y)) return;
+	/* ...but keep coordinates relative to the window. */
+	new_dpi.left += this->pos_x;
+	new_dpi.top += this->pos_y;
+
+	DrawPixelInfo *old_dpi = _cur_dpi;
+	_cur_dpi = &new_dpi;
+
 	Rect r;
 	r.left = this->pos_x;
 	r.right = this->pos_x + this->current_x - 1;
 	r.top = this->pos_y;
 	r.bottom = this->pos_y + this->current_y - 1;
 
-	const DrawPixelInfo *dpi = _cur_dpi;
-	if (dpi->left > r.right || dpi->left + dpi->width <= r.left || dpi->top > r.bottom || dpi->top + dpi->height <= r.top) return;
-
 	bool clicked = this->IsLowered();
 	switch (this->type) {
 		case WWT_EMPTY:
@@ -2512,6 +2519,8 @@
 	if (this->IsDisabled()) {
 		GfxFillRect(r.left + 1, r.top + 1, r.right - 1, r.bottom - 1, _colour_gradient[this->colour & 0xF][2], FILLRECT_CHECKER);
 	}
+
+	_cur_dpi = old_dpi;
 }
 
 /**
